source/extensions/filters/http/grpc_stats/grpc_stats_filter.cc

```diff
@@ -17,66 +19,201 @@ namespace GrpcStats {
+                                     const envoy::config::core::v3::GrpcMethodList& method_list)
+      : stat_name_pool_(symbol_table), map_(populate(method_list)) {}
+
+  absl::optional<Grpc::Context::RequestStatNames>
+  lookup(const Grpc::Common::RequestNames& request_names) const {
+    auto it = map_.find(request_names);
+    if (it != map_.end()) {
+      return it->second;
+    }
+
+    return {};
+  }
+
+private:
+  using OwningKey = std::tuple<std::string, std::string>;
+  using ViewKey = Grpc::Common::RequestNames;
+
+  class MapHash {
+  private:
+    // Use the same type for hashing all variations to ensure the same hash value from all source
+    // types.
+    using ViewTuple = std::tuple<absl::string_view, absl::string_view>;
+    static uint64_t hash(const ViewTuple& key) { return absl::Hash<ViewTuple>()(key); }
+
+  public:
+    using is_transparent = void;
+
+    uint64_t operator()(const OwningKey& key) const { return hash(key); }
+    uint64_t operator()(const ViewKey& key) const {
+      return hash(ViewTuple(key.service_, key.method_));
+    }
+  };
+
+  struct MapEq {
+    using is_transparent = void;
+    bool operator()(const OwningKey& left, const OwningKey& right) const { return left == right; }
+    bool operator()(const OwningKey& left, const ViewKey& right) const {
+      return left == std::make_tuple(right.service_, right.method_);
+    }
+  };
+  using MapType = absl::flat_hash_map<OwningKey, Grpc::Context::RequestStatNames, MapHash, MapEq>;
+
+  // Helper for generating a populated MapType so that `map_` can be const.
+  MapType populate(const envoy::config::core::v3::GrpcMethodList& method_list) {
+    MapType map;
+    for (const auto& service : method_list.services()) {
+      Stats::StatName stat_name_service = stat_name_pool_.add(service.name());
+
+      for (const auto& method_name : service.method_names()) {
+        Stats::StatName stat_name_method = stat_name_pool_.add(method_name);
+        map[OwningKey(service.name(), method_name)] =
+            Grpc::Context::RequestStatNames{stat_name_service, stat_name_method};
+      }
+    }
+    return map;
+  }
+
+  Stats::StatNamePool stat_name_pool_;
+  const MapType map_;
+};
+
+struct Config {
+  Config(const envoy::extensions::filters::http::grpc_stats::v3::FilterConfig& proto_config,
+         Server::Configuration::FactoryContext& context)
+      : context_(context.grpcContext()), emit_filter_state_(proto_config.emit_filter_state()) {
+
+    switch (proto_config.per_method_stat_specifier_case()) {
+    case envoy::extensions::filters::http::grpc_stats::v3::FilterConfig::
+        PER_METHOD_STAT_SPECIFIER_NOT_SET:
+    case envoy::extensions::filters::http::grpc_stats::v3::FilterConfig::kStatsForAllMethods:
+      stats_for_all_methods_ =
+          PROTOBUF_GET_WRAPPED_OR_DEFAULT(proto_config, stats_for_all_methods,
+                                          context.runtime().snapshot().deprecatedFeatureEnabled(
```

Comment:
@ggreenway  Quick drive be here: I'm pretty sure there is not going to print a warning in the default case, right? (It will inc the stat I think.) This is going to make it hard/impossible for operators to figure what they did wrong. If I'm right can we do a follow up to get a warning printed somehow? cc @zuercher @alyssawilk who have more experience with the deprecation warning code.

Reply:
Good catch; I'll sort out a fix.